pkg/controller/installation/bpf_test.go

```diff
@@ -0,0 +1,523 @@
+			fc := &crdv1.FelixConfiguration{ObjectMeta: metav1.ObjectMeta{Name: "default", Annotations: fcAnnotations}}
+			Expect(c.Create(ctx, fc)).NotTo(HaveOccurred())
+			Expect(updateBPFEnabledAllowed(fc)).Should(HaveOccurred())
+		})
+
+		It("should query FelixConfig annotation is set and spec is set and matches then outcome is valid", func() {
+			bpfEnabled := true
+			fcAnnotations := make(map[string]string)
+			fcAnnotations[render.BPFOperatorAnnotation] = "true"
+			fc := &crdv1.FelixConfiguration{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:        "default",
+					Annotations: fcAnnotations,
+				},
+				Spec: crdv1.FelixConfigurationSpec{
+					BPFEnabled: &bpfEnabled,
+				},
+			}
+			Expect(c.Create(ctx, fc)).NotTo(HaveOccurred())
+			Expect(updateBPFEnabledAllowed(fc)).ShouldNot(HaveOccurred())
+		})
+
+		It("should query FelixConfig annotation is set and spec is set but does not match then outcome is invalid", func() {
+			bpfEnabled := false
+			fcAnnotations := make(map[string]string)
+			fcAnnotations[render.BPFOperatorAnnotation] = "true"
+			fc := &crdv1.FelixConfiguration{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:        "default",
+					Annotations: fcAnnotations,
+				},
+				Spec: crdv1.FelixConfigurationSpec{
+					BPFEnabled: &bpfEnabled,
+				},
+			}
+			Expect(c.Create(ctx, fc)).NotTo(HaveOccurred())
+			Expect(updateBPFEnabledAllowed(fc)).Should(HaveOccurred())
+		})
+
+		It("should query calico-node DS in BPF dataplane and if DS status not set then verify rollout not complete", func() {
+			// Arrange.
+			// Upgrade cluster from IP Tables to BPF dataplane.
+			cr := createInstallation(c, ctx, operator.LinuxDataplaneBPF)
+
+			// Create calico-node Daemonset annotation to indicate update rollout complete.
+			container := corev1.Container{Name: common.NodeDaemonSetName}
+			ds := &appsv1.DaemonSet{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      common.NodeDaemonSetName,
+					Namespace: common.CalicoNamespace,
+				},
+				Spec: appsv1.DaemonSetSpec{
+					Template: corev1.PodTemplateSpec{
+						ObjectMeta: metav1.ObjectMeta{},
+						Spec:       corev1.PodSpec{Containers: []corev1.Container{container}},
+					},
+				},
+			}
+			Expect(c.Create(ctx, ds)).NotTo(HaveOccurred())
+
+			// Create felix config
+			fc := &crdv1.FelixConfiguration{ObjectMeta: metav1.ObjectMeta{Name: "default"}}
+			Expect(c.Create(ctx, fc)).NotTo(HaveOccurred())
+
+			// Act.
+			_, err := utils.PatchFelixConfiguration(ctx, r.client, func(fc *crdv1.FelixConfiguration) bool {
+				return r.setBPFUpdatesOnFelixConfiguration(cr, ds, fc, reqLogger)
+			})
+			Expect(err).ShouldNot(HaveOccurred())
+
+			// Assert.
+			err = c.Get(ctx, types.NamespacedName{Name: "default"}, fc)
+			Expect(err).ShouldNot(HaveOccurred())
+			Expect(fc.Spec.BPFEnabled).To(BeNil())
+		})
+
+		It("should query calico-node DS in BPF dataplane and if DS status rolling out then verify rollout not complete", func() {
```

Comment:
This test is also largely testing `PatchFelixConfiguration` - if we just want to test `setBPFUpdatesOnFelixConfiguration` here, we don't need to create the installation, we don't need to create the DS, we don't need to create the FelixConfig (in fact the fake client isn't needed at all).

Like I implied in the other comment, testing for `PatchFelixConfiguration` belongs in the `utils` package. The tests here should primarily be around the functions in this package.